src/TypeSystem/Resolver/TernaryOperation/TernaryOperationTypeResolver.php

```diff
@@ -41,14 +44,39 @@ public function resolveTypeOf(TernaryOperationNode $ternaryOperationNode): TypeI
         $expressionTypeResolver = new ExpressionTypeResolver(
             scope: $this->scope
         );
-        $conditionNode = $ternaryOperationNode->condition->root;
+        $conditionNode = $ternaryOperationNode->condition;
 
-        if ($conditionNode instanceof BooleanLiteralNode) {
-            return $conditionNode->value
+        $rootType = $expressionTypeResolver->resolveTypeOf($conditionNode);
+
+        if ($conditionNode->root instanceof BooleanLiteralNode) {
+            return $conditionNode->root->value
                 ? $expressionTypeResolver->resolveTypeOf($ternaryOperationNode->true)
                 : $expressionTypeResolver->resolveTypeOf($ternaryOperationNode->false);
         }
 
+        if ($conditionNode->root instanceof IdentifierNode && $rootType instanceof UnionType && $rootType->isNullable()) {
+            $trueExpressionTypeResolver = new ExpressionTypeResolver(
+                scope: new ShallowScope(
+                    $conditionNode->root->value,
+                    $rootType->withoutNullable(),
+                    $this->scope
+                )
+            );
+
+            $falseExpressionTypeResolver = new ExpressionTypeResolver(
+                scope: new ShallowScope(
+                    $conditionNode->root->value,
+                    NullType::get(),
+                    $this->scope
+                )
+            );
+
+            return UnionType::of(
+                $trueExpressionTypeResolver->resolveTypeOf($ternaryOperationNode->true),
+                $falseExpressionTypeResolver->resolveTypeOf($ternaryOperationNode->false)
+            );
+        }
+
```

Comment:
Implemented :D But now im too tired to type things down that i encountered, see you tomorrow.

(And thanks for the tipp with TernaryBranchScope works like a charm and even better ^^)

Also i cant thank you enough for your most precise code reviews and comments - they are extremely helpful and well guiding me! Thank you really so much!!!

Reply:
yes youre right, my implementation is only limited for super simple things like `foo ? foo : ""`

i thought about how we can i implement a universal solution, which also handles any thinkable more complex case, but it seems like a hard challenge

for example given `foo` is a nullable string.

those expressions will be resolved as of now correctly:

```
foo ? foo : ""
foo !== null ? foo : ""
```

but here we loose it:

```
(foo) ? foo : ""
foo !== nullVariable ? foo : ""
((foo)) ? foo : ""
(foo !== null) ? foo : ""
foo && true ? foo : ""
(foo ? foo : foo) ? foo : ""
((foo ? foo : foo) ? foo : foo) ? foo : ""
```

I checked all the examples with typescript, and it does a good job (unsurprisingly). I even went as far as trying to have a look at the ts codebase for some inspiration, but i couldnt even find the right unit tests to start with xD

Some of the above are surely not a real world use case but its a simple demonstration how easy it is to confuse the implemented type resolution (or type inference?).

I think i need to find out while resolving the condition part, which variable participated and how it affected the truthyness. Im not really sure if this would be the right approach in the long run and if it can be easily implemented.

But for now im also quite happy with my current implementation as a start.

Reply:
Found the place where phpstan does it:
https://github.com/phpstan/phpstan-src/blob/07bb4aa2d5e39dafa78f56c5df132c763c2d1b67/src/Analyser/MutatingScope.php#L1616

```php
// $node is a TernaryExpression with the fields: cond, if and else

$booleanConditionType = $this->getType($node->cond)->toBoolean();
if ($booleanConditionType->isTrue()->yes()) {
     return $this->filterByTruthyValue($node->cond)->getType($node->if);
}

if ($booleanConditionType->isFalse()->yes()) {
    return $this->filterByFalseyValue($node->cond)->getType($node->else);
}

return TypeCombinator::union(
     $this->filterByTruthyValue($node->cond)->getType($node->if),
     $this->filterByFalseyValue($node->cond)->getType($node->else),
);
```